src/ConfigurationClientManager.ts

```diff
@@ -0,0 +1,310 @@
+    }
+
+    async getClients() : Promise<ConfigurationClientWrapper[]> {
+        if (!this.isFailoverable) {
+            return this.#staticClients;
+        }
+
+        const currentTime = Date.now();
+        if (this.#isFallbackClientDiscoveryDue(currentTime)) {
+            this.#lastFallbackClientRefreshAttempt = currentTime;
+            const host = new URL(this.endpoint).hostname;
+            await this.#discoverFallbackClients(host);
+        }
+
+        // Filter static clients whose backoff time has ended
+        let availableClients = this.#staticClients.filter(client => client.backoffEndTime <= currentTime);
+        // If there are dynamic clients, filter and concatenate them
+        if (this.#dynamicClients && this.#dynamicClients.length > 0) {
+            availableClients = availableClients.concat(
+                this.#dynamicClients
+                    .filter(client => client.backoffEndTime <= currentTime));
+        }
+
+        return availableClients;
+    }
+
+    async refreshClients() {
+        const currentTime = Date.now();
+        if (this.isFailoverable &&
+            currentTime > new Date(this.#lastFallbackClientRefreshAttempt + MINIMAL_CLIENT_REFRESH_INTERVAL).getTime()) {
+            this.#lastFallbackClientRefreshAttempt = currentTime;
+            const host = new URL(this.endpoint).hostname;
+            await this.#discoverFallbackClients(host);
+        }
+    }
+
+    async #discoverFallbackClients(host) {
+        const timeout = setTimeout(() => {
+        }, SRV_QUERY_TIMEOUT);
+        const srvResults = await querySrvTargetHost(host);
+
+        try {
+            const result = await Promise.race([srvResults, timeout]);
+
+            if (result === timeout) {
+                throw new Error("SRV record query timed out.");
+            }
+
+            const srvTargetHosts = result as string[];
+            // Shuffle the list of SRV target hosts
+            for (let i = srvTargetHosts.length - 1; i > 0; i--) {
+                const j = Math.floor(Math.random() * (i + 1));
+                [srvTargetHosts[i], srvTargetHosts[j]] = [srvTargetHosts[j], srvTargetHosts[i]];
+            }
+
+            const newDynamicClients: ConfigurationClientWrapper[] = [];
+            for (const host of srvTargetHosts) {
+                if (isValidEndpoint(host, this.#validDomain)) {
+                    const targetEndpoint = `https://${host}`;
+                    if (targetEndpoint.toLowerCase() === this.endpoint.toLowerCase()) {
+                        continue;
+                    }
+                    const client = this.#newConfigurationClient(targetEndpoint);
+                    newDynamicClients.push(new ConfigurationClientWrapper(targetEndpoint, client));
+                }
+            }
+
+            this.#dynamicClients = newDynamicClients;
+            this.#lastFallbackClientRefreshTime = Date.now();
+        } catch (err) {
```

Comment:
We should only allow querySrv to fail. Most of the code in try {} should be put outside.